include/s3select_functions.h

```diff
@@ -643,226 +774,305 @@ struct _fn_to_timestamp : public base_function
+
+    new_ptime = *val_timestamp.timestamp();
+    result->set_value( (int64_t)new_ptime.date().year());
+    return true;
+  }
+};
+
+struct _fn_extract_month_from_timestamp : public base_date_extract
+{
+  boost::posix_time::ptime new_ptime;
 
   bool operator()(bs_stmt_vec_t* args, variable* result) override
   {
-    auto iter = args->begin();
-    int args_size = args->size();
+    param_validation(args);
 
-    if (args_size < 2)
-    {
-      throw base_s3select_exception("to_timestamp should have 2 parameters");
-    }
+    new_ptime = *val_timestamp.timestamp();
+    result->set_value( (int64_t)new_ptime.date().month());
+    return true;
+  }
+};
 
-    base_statement* date_part = *iter;
+struct _fn_extract_day_from_timestamp : public base_date_extract
+{
+  boost::posix_time::ptime new_ptime;
 
-    val_date_part = date_part->eval();//TODO could be done once?
+  bool operator()(bs_stmt_vec_t* args, variable* result) override
+  {
+    param_validation(args);
 
-    if(val_date_part.is_string()== false)
-    {
-      throw base_s3select_exception("first parameter should be string");
-    }
+    new_ptime = *val_timestamp.timestamp();
+    result->set_value( (int64_t)new_ptime.date().day_of_year());
+    return true;
+  }
+};
 
-    iter++;
+struct _fn_extract_hour_from_timestamp : public base_date_extract
+{
+  boost::posix_time::ptime new_ptime;
 
-    base_statement* ts = *iter;
+  bool operator()(bs_stmt_vec_t* args, variable* result) override
+  {
+    param_validation(args);
 
-    if(ts->eval().is_timestamp()== false)
-    {
-      throw base_s3select_exception("second parameter is not timestamp");
-    }
+    new_ptime = *val_timestamp.timestamp();
+    result->set_value( (int64_t)new_ptime.time_of_day().hours());
+    return true;
+  }
+};
 
-    new_ptime = *ts->eval().timestamp();
+struct _fn_extract_minute_from_timestamp : public base_date_extract
+{
+  boost::posix_time::ptime new_ptime;
 
-    if( strcmp(val_date_part.str(), "year")==0 )
-    {
-      result->set_value( (int64_t)new_ptime.date().year() );
-    }
-    else if( strcmp(val_date_part.str(), "month")==0 )
-    {
-      result->set_value( (int64_t)new_ptime.date().month() );
-    }
-    else if( strcmp(val_date_part.str(), "day")==0 )
-    {
-      result->set_value( (int64_t)new_ptime.date().day_of_year() );
-    }
-    else if( strcmp(val_date_part.str(), "week")==0 )
-    {
-      result->set_value( (int64_t)new_ptime.date().week_number() );
-    }
-    else
-    {
-      throw base_s3select_exception(std::string( val_date_part.str() + std::string("  is not supported ") ).c_str() );
-    }
+  bool operator()(bs_stmt_vec_t* args, variable* result) override
+  {
+    param_validation(args);
 
+    new_ptime = *val_timestamp.timestamp();
+    result->set_value( (int64_t)new_ptime.time_of_day().minutes());
     return true;
   }
+};
 
+struct _fn_extract_second_from_timestamp : public base_date_extract
+{
+  boost::posix_time::ptime new_ptime;
```

Comment:
/boost::posix_time::ptime new_ptime;/ ( and new_ptime = *val_timestamp.timestamp() )
could be a part of the base class